go-controller/pkg/ovn/master.go

```diff
@@ -174,56 +194,73 @@ func (oc *Controller) SetupMaster(masterNodeName string) error {
-		joinSubnet = "fd98::1/64"
-	} else {
-		joinSubnet = "100.64.0.1/16"
+// annotate the node with the join subnet information assigned to node's join logical switch.
+// the format of the annotation is:
+//
+// k8s.ovn.org/node-join-subnets: {
+//	  "default": "100.64.0.0/29",
+// }
+func (oc *Controller) addNodeJoinSubnetAnnotations(node *kapi.Node, subnet string) error {
+	// nothing to do if the node already has the annotation key
+	_, ok := node.Annotations[OvnNodeJoinSubnets]
+	if ok {
+		return nil
 	}
-	joinIP, joinCIDR, _ := net.ParseCIDR(joinSubnet)
-	if config.IPv6Mode {
-		stdout, stderr, err = util.RunOVNNbctl("--may-exist", "ls-add", "join",
-			"--", "set", "logical_switch", "join", fmt.Sprintf("%s=%s", config.OtherConfigSubnet(), joinCIDR.String()))
-	} else {
-		stdout, stderr, err = util.RunOVNNbctl("--may-exist", "ls-add", "join",
-			"--", "set", "logical_switch", "join", fmt.Sprintf("%s=%s", config.OtherConfigSubnet(), joinCIDR.String()),
-			"--", "set", "logical_switch", "join", fmt.Sprintf("other-config:exclude_ips=%s", joinIP.String()))
+
+	bytes, err := json.Marshal(map[string]string{
+		"default": subnet,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to marshal node %q annotation %q for subnet %s",
+			node.Name, OvnNodeJoinSubnets, subnet)
 	}
+	nodeAnnotations := make(map[string]interface{})
+	nodeAnnotations[OvnNodeJoinSubnets] = string(bytes)
+	err = oc.kube.SetAnnotationsOnNode(node, nodeAnnotations)
 	if err != nil {
-		logrus.Errorf("Failed to create logical switch called \"join\", stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
-		return err
+		return fmt.Errorf("failed to set node annotation %q on existing node %s to %q: %v",
+			OvnNodeJoinSubnets, node.Name, subnet, err)
 	}
+	return nil
+}
 
-	// Connect the distributed router to "join".
-	routerMac, stderr, err := util.RunOVNNbctl("--if-exist", "get", "logical_router_port", "rtoj-"+clusterRouter, "mac")
+func (oc *Controller) allocateJoinSubnet(node *kapi.Node) (string, error) {
+	joinSubnet, _ := parseNodeJoinSubnet(node)
+	if joinSubnet != nil {
+		return joinSubnet.String(), nil
+	}
+
+	// Allocate a new network for the join switch
+	joinSubnetStr, err := oc.joinSubnetAllocator.AllocateNetwork()
 	if err != nil {
-		logrus.Errorf("Failed to get logical router port rtoj-%v, stderr: %q, error: %v", clusterRouter, stderr, err)
-		return err
+		return "", fmt.Errorf("Error allocating subnet for join switch for node  %s: %v", node.Name, err)
 	}
-	if routerMac == "" {
-		routerMac = util.GenerateMac()
-		stdout, stderr, err = util.RunOVNNbctl("--", "--may-exist", "lrp-add", clusterRouter,
-			"rtoj-"+clusterRouter, routerMac, joinSubnet)
+
+	defer func() {
+		// Release the allocation on error
 		if err != nil {
-			logrus.Errorf("Failed to add logical router port rtoj-%v, stdout: %q, stderr: %q, error: %v",
-				clusterRouter, stdout, stderr, err)
-			return err
+			_ = oc.joinSubnetAllocator.ReleaseNetwork(joinSubnetStr)
```

Comment:
@trozet this is in defer function. You will not have context of failure from the defer function since it can be called from any of the error exit points in the function. I think what I have now mimics the NodeSubnetAllocator code.